src/szk.rs

```diff
@@ -0,0 +1,459 @@
+//! Szk is a wrapper class for the FLP proof system (accessible through the `Type` trait)
+//! that is designed to accept inputs that are already secret-shared between two aggregators
+//! and returns proofs that are also shared between the two aggregators. The underlying FLP
+//! accepts inputs that are vectors over a finite field, and its random coins and proofs are also
+//! vectors over the same field. In contrast, only the leader's SZK proof share contains any finite
+//! field elements, and the initial coins are bytestrings (`Seed`s). The wrapper class defined
+//! here uses an Xof (to be modeled as a random oracle) to sample coins and the helper's proof share,
+//! following the strategy of Prio3.
+
+use crate::codec::{Encode, CodecError};
+use crate::field::{FieldElement, FftFriendlyFieldElement};
+use crate::flp::{Type, FlpError};
+use crate::vdaf::xof::{Seed, Xof, IntoFieldVec, XofTurboShake128};
+#[cfg(feature = "experimental")]
+use crate::prng::{Prng, PrngError};
+use std::marker::PhantomData;
+///Domain separation tags
+const DST_PROVE_RANDOMNESS: u16 = 0;
+const DST_PROOF_SHARE: u16 = 1;
+const DST_QUERY_RANDOMNESS: u16 = 2;
+const DST_JOINT_RAND_SEED: u16 = 3;
+const DST_JOINT_RAND_PART: u16 = 4;
+const DST_JOINT_RANDOMNESS: u16 = 5;
+
+/// Errors propagated by methods in this module.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum SzkError {
+    /// Calling [`Type::prove`] returned an error.
+    #[error("prove error: {0}")]
+    Prove(String),
+
+    /// Calling [`Type::query`] returned an error.
+    #[error("query error: {0}")]
+    Query(String),
+
+    /// Calling [`Type::decide`] returned an error.
+    #[error("decide error: {0}")]
+    Decide(String),
+
+    /// Generic invalid parameter. This may be returned when a type cannot be constructed.
+    #[error("invalid paramter: {0}")]
+    InvalidParameter(String),
+
+    /// Returned if an FLP operation encountered an error.
+    #[error("Flp error: {0}")]
+    Field(#[from] FlpError),
+
+    /// PRNG error.
+    #[error("prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// PRNG error.
+    #[error("codec error: {0}")]
+    Codec(#[from] CodecError),
+
+}
+
+/// Contains an FLP proof share, and if joint randomness is needed, the blind
+/// used to derive it and the other party's joint randomness part.
+pub enum SzkProofShare<'a, const SEED_SIZE: usize, F: FftFriendlyFieldElement>{
+///Leader's proof share is uncompressed.
+///The first Seed is a blind, second is a joint randomness part.
+Leader(Vec<F>, Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>),
+///Helper's seed is compressed using the same blind as its joint randomness
+Helper(Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>)
+}
+
+/// Verifier type for Szk Proof
+pub type SzkVerifier<F> = Vec<F>;
+
+
+///Main struct encapsulating the shared zero-knowledge functionality.
+/// The type T is the undelrying FLP proof system
```

Comment:
nit: these sentences would run together after Markdown processing
```suggestion
/// The type T is the underlying FLP proof system.
```